ao-tools/ao-load/ao-load.1
.\"
.\" Copyright © 2009 Keith Packard <keithp@keithp.com>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
.\"
.\"
.TH AO-LOAD 1 "ao-load" ""
.SH NAME
ao-load \- flash a program to a AltOS device
.SH SYNOPSIS
.B "ao-load"
[\-T \fItty-device\fP]
[\--tty \fItty-device\fP]
[\-D \fIaltos-device\fP]
[\--device \fIaltos-device\fP]
[\--cal \fIradio-calibration\fP]
\fIfile.ihx\fP
\fIdevice serial number\fP
.SH DESCRIPTION
.I ao-load
loads the specified .ihx file into the target device flash memory,
customizing the AltOS image with the specified serial number.
.SH OPTIONS
.TP
\-T tty-device | --tty tty-device
This selects which tty device the debugger uses to communicate with
the target device. The special name 'BITBANG' directs ao-dbg to use
the cp2103 connection, otherwise this should be a usb serial port
connected to a suitable cc1111 debug node.
.TP
\-D AltOS-device | --device AltOS-device
Search for a connected device. This requires an argument of one of the
following forms:
.IP
TeleMetrum:2
.br
TeleMetrum
.br
2
.IP
Leaving out the product name will cause the tool to select a suitable
product, leaving out the serial number will cause the tool to match
one of the available devices.
.TP
\-c radio-calibration | --cal radio-calibration
This programs the radio calibration value into the image for hardware
which doesn't have any eeprom storage for this value. The value here
can be computed given the current radio calibration value, the
measured frequency and the desired frequency:
.IP
	cal' = cal * (desired/measured)
.IP
The default calibration value is 1186611.
.SH USAGE
.I ao-load
reads the specified .ihx file into memory, locates the matching .map
file and edits the image to customize it using the specified serial
number. It then connects to the specified target device and writes the
program to the target device flash memory.
.SH AUTHOR
Keith Packard


ChangeLog
commit 10d1bbcd9709a5eee8d50989215242b16feb7232
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Dec 19 11:34:16 2009 -0800

    Use ao_radio_get/ao_radio_put in packet code.
    
    The ao_radio_get function both acquires the mutex *and* configures the
    radio channel and frequency. Failing to use this in the packet code
    would leave the radio frequency unconfigured.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit a15abc1882a3bdd2c980eed169f3b80337528390
Author: Bdale Garbee <bdale@gag.com>
Date:   Sat Dec 19 12:15:29 2009 -0700

    update changelogs for Debian build

commit bbb152c712801653374a8f82869e2e8bf41f7279
Author: Bdale Garbee <bdale@gag.com>
Date:   Sat Dec 19 12:05:40 2009 -0700

    update changelogs for Debian build

commit 7ac8efea5d60b81adccdc1e38ac4c13facfae7c9
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Dec 15 23:58:29 2009 -0800

    Add --cal to man page

commit 5481082b18226a0de6b377215b3b330bdbc4a6c6
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Dec 5 19:50:38 2009 -0800

    Allow radio calibration to be set from ao-load
    
    This moves the initial radio calibration value into const memory where
    it will be used if no eeprom configuration value is available, either
    on an unprogrammed board with eeprom or a device without an eeprom.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 46f03ab3145a61139c8ca6fc99e8f2798728b5a9
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Dec 5 15:36:12 2009 -0800

    Re-order config values. Change frequency to cal
    
    Place more often used values at top, and consistently call the radio
    value 'calibration' instead of 'frequency'.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit bf29a62532fec12e6af2d2f3a6624882c863e933
Author: Bdale Garbee <bdale@gag.com>
Date:   Sat Dec 5 11:03:42 2009 -0700

    update changelogs for Debian build

commit cd49847f3125df1733f298b56a43e8027ab5ce05
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Dec 4 23:38:26 2009 -0800

    Add radio calibration configuration.
    
    The crystal we use is only good for 20ppm, which generates a fairly
    significant error bounds at our RF frequency. This commit adds a
    configuration variable that sets the RF frequency control variable so
    that the output frequency can be adjusted.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit c671a3f31f54715284beef5baa4a72ca922e4018
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Dec 4 23:36:54 2009 -0800

    Make ao_cmd_decimal produce both 32 and 16 bit values.
    
    ao_cmd_lex_u32 is produced in addition to ao_cmd_lex_i so that
    functions can easily read 32-bit values from the command line.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit e6d6ad66ce177eb6e49eaa04b1ecc6426f4bbdbf
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Dec 4 23:35:40 2009 -0800

    Remove send_serial and serial_baud commands.
    
    No longer useful, and they take up space.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit ce7325cefbbe91054c1e8174cf40a6c687f6694b
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Dec 4 21:02:07 2009 -0700

    update changelogs for Debian build

commit 9a1d7dd78c07d18954ef620b6ed25d6701df21e2
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Dec 4 14:35:39 2009 -0800

    Change default callsign to "N0CALL"

commit 291e2004e77debdc1543912cf11043c849305de8
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Nov 26 11:18:37 2009 -0700

    update changelogs for Debian build

commit 40533095a96b6c6364eebdc4b7d53c4eabe72e9a
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Nov 26 11:16:04 2009 -0700

    update changelogs for Debian build

commit ab1075e7219b02258c1613d93379582be4168947
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Nov 26 11:15:13 2009 -0700

    change home URL in control file to be the AltOS page

commit d34dcc5f3616e59ee90ed172770fe2a3eb1e0cac
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Nov 26 10:40:07 2009 -0700

    fix absolute path in debian/dirs, add Suggests for slim-altusmetrum

commit ea40561b36519a5dcabedabe18672b79ea5e9993
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Nov 22 10:18:44 2009 -0700

    update changelogs for Debian build

commit d6ba07e885bdc62ba64719c9d8cc42fcecbcb09d
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Nov 22 01:10:44 2009 -0800

    Automatically extract flight number for eeprom and telem filenames.
    
    Extract flight number from either telemetry or eeprom files and use
    that in the resulting filenames. To ensure that files remain unique,
    add a new field, -seq-%03d. This is appended only when the sequence
    number is non-zero as it shouldn't occur in normal usage.
    
    This also eliminates some duplicate filename creation code in the
    library and aoview sources.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 06cebd1026dc1bd6ee51526fa2d02905df3b3b37
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Nov 22 00:52:58 2009 -0800

    ao-postflight: don't try to use missing gps sat data
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit a9ada1b538af3308e1b22bd024d9204521184173
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Nov 21 22:12:21 2009 -0800

    ao-postflight: compute barometric alt for each GPS position
    
    Print that to the --gps file, and use that in the --kml file for the
    altitude. Gives a very different picture of our flight tracks,
    presumably far more accurate (at least in altitude).
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit b84b634d9ae8ce6ab1c02833a3ed8514404e1ca3
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Nov 21 22:11:45 2009 -0800

    Don't crash if --plot isn't passed on ao-postflight command line
    
    Crashing is not nice.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 2d77c18b15834046b7b79d49d87211828f2409e9

ao-tools/ao-load/ao-load.c
/*
 * Copyright © 2008 Keith Packard <keithp@keithp.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 */

#include <stdlib.h>
#include <limits.h>
#include <stdint.h>
#include <unistd.h>
#include <getopt.h>
#include "ccdbg.h"
#include "cc.h"

#define AO_USB_DESC_STRING		3

struct sym {
	unsigned	addr;
	char		*name;
} ao_symbols[] = {
	{ 0,	"_ao_serial_number" },
#define AO_SERIAL_NUMBER	(ao_symbols[0].addr)
	{ 0,	"_ao_usb_descriptors" },
#define AO_USB_DESCRIPTORS	(ao_symbols[1].addr)
	{ 0,	"_ao_radio_cal" },
#define AO_RADIO_CAL		(ao_symbols[2].addr)
};

#define NUM_SERIAL_SYMBOLS	2
#define NUM_SYMBOLS		3

static int
find_symbols(FILE *map)
{
	char	line[2048];
	char	*addr, *addr_end;
	char	*name;
	char	*save;
	char	*colon;
	unsigned long	a;
	int	s;
	int	found = 0;

	while (fgets(line, sizeof(line), map) != NULL) {
		line[sizeof(line)-1] = '\0';
		addr = strtok_r(line, " \t\n", &save);
		if (!addr)
			continue;
		name = strtok_r(NULL, " \t\n", &save);
		if (!name)
			continue;
		colon = strchr (addr, ':');
		if (!colon)
			continue;
		a = strtoul(colon+1, &addr_end, 16);
		if (a == ULONG_MAX || addr_end == addr)
			continue;
		for (s = 0; s < NUM_SYMBOLS; s++)
			if (!strcmp(ao_symbols[s].name, name)) {
				ao_symbols[s].addr = (unsigned) a;
				++found;
				break;
			}
	}
	return found >= NUM_SERIAL_SYMBOLS;
}

static int
rewrite(struct hex_image *image, unsigned addr, char *data, int len)
{
	int i;
	if (addr < image->address || image->address + image->length < addr + len)
		return 0;
	printf("rewrite %04x:", addr);
	for (i = 0; i < len; i++)
		printf (" %02x", image->data[addr - image->address + i]);
	printf(" ->");
	for (i = 0; i < len; i++)
		printf (" %02x", data[i]);
	printf("\n");
	memcpy(image->data + addr - image->address, data, len);
}

static const struct option options[] = {
	{ .name = "tty", .has_arg = 1, .val = 'T' },
	{ .name = "device", .has_arg = 1, .val = 'D' },
	{ .name = "cal", .has_arg = 1, .val = 'c' },
	{ 0, 0, 0, 0},
};

static void usage(char *program)
{
	fprintf(stderr, "usage: %s [--tty <tty-name>] [--device <device-name>] [--cal <radio-cal>] file.ihx serial-number\n", program);
	exit(1);
}

int
main (int argc, char **argv)
{
	struct ccdbg	*dbg;
	uint8_t		status;
	uint16_t	pc;
	struct hex_file	*hex;
	struct hex_image *image;
	char		*filename;
	FILE		*file;
	FILE		*map;
	char		*serial_string;
	unsigned int	serial;
	char		*mapname, *dot;
	char		*serial_ucs2;
	int		serial_ucs2_len;
	char		serial_int[2];
	unsigned int	s;
	int		i;
	unsigned	usb_descriptors;
	int		string_num;
	char		*tty = NULL;
	char		*device = NULL;
	uint32_t	cal = 0;
	char		cal_int[4];
	char		*cal_end;
	int		c;

	while ((c = getopt_long(argc, argv, "T:D:c:", options, NULL)) != -1) {
		switch (c) {
		case 'T':
			tty = optarg;
			break;
		case 'D':
			device = optarg;
			break;
		case 'c':
			cal = strtoul(optarg, &cal_end, 10);
			if (cal_end == optarg || *cal_end != '\0')
				usage(argv[0]);
			break;
		default:
			usage(argv[0]);
			break;
		}
	}
	filename = argv[optind];
	if (filename == NULL)
		usage(argv[0]);
	mapname = strdup(filename);
	dot = strrchr(mapname, '.');
	if (!dot || strcmp(dot, ".ihx") != 0)
		usage(argv[0]);
	strcpy(dot, ".map");

	serial_string = argv[optind + 1];
	if (serial_string == NULL)
		usage(argv[0]);

	file = fopen(filename, "r");
	if (!file) {
		perror(filename);
		exit(1);
	}
	map = fopen(mapname, "r");
	if (!map) {
		perror(mapname);
		exit(1);
	}
	if (!find_symbols(map)) {
		fprintf(stderr, "Cannot find symbols in \"%s\"\n", mapname);
		exit(1);
	}
	fclose(map);

	hex = ccdbg_hex_file_read(file, filename);
	fclose(file);
	if (!hex) {
		perror(filename);
		exit (1);
	}
	image = ccdbg_hex_image_create(hex);
	if (!image) {
		fprintf(stderr, "image create failed\n");
		exit (1);
	}
	ccdbg_hex_file_free(hex);

	serial = strtoul(serial_string, NULL, 0);
	if (!serial)
(argv[0]);

	serial_int[0] = serial & 0xff;
	serial_int[1] = (serial >> 8) & 0xff;

	if (!rewrite(image, AO_SERIAL_NUMBER, serial_int, sizeof (serial_int))) {
		fprintf(stderr, "Cannot rewrite serial integer at %04x\n",
			AO_SERIAL_NUMBER);
		exit(1);
	}

	usb_descriptors = AO_USB_DESCRIPTORS - image->address;
	string_num = 0;
	while (image->data[usb_descriptors] != 0 && usb_descriptors < image->length) {
		if (image->data[usb_descriptors+1] == AO_USB_DESC_STRING) {
			++string_num;
			if (string_num == 4)
				break;
		}
		usb_descriptors += image->data[usb_descriptors];
	}
	if (usb_descriptors >= image->length || image->data[usb_descriptors] == 0 ) {
		fprintf(stderr, "Cannot rewrite serial string at %04x\n", AO_USB_DESCRIPTORS);
		exit(1);
	}

	serial_ucs2_len = image->data[usb_descriptors] - 2;
	serial_ucs2 = malloc(serial_ucs2_len);
	if (!serial_ucs2) {
		fprintf(stderr, "Malloc(%d) failed\n", serial_ucs2_len);
		exit(1);
	}
	s = serial;
	for (i = serial_ucs2_len / 2; i; i--) {
		serial_ucs2[i * 2 - 1] = 0;
		serial_ucs2[i * 2 - 2] = (s % 10) + '0';
		s /= 10;
	}
	if (!rewrite(image, usb_descriptors + 2 + image->address, serial_ucs2, serial_ucs2_len))
		usage(argv[0]);

	if (cal) {
		cal_int[0] = cal & 0xff;
		cal_int[1] = (cal >> 8) & 0xff;
		cal_int[2] = (cal >> 16) & 0xff;
		cal_int[3] = (cal >> 24) & 0xff;
		if (!AO_RADIO_CAL) {
			fprintf(stderr, "Cannot find radio calibration location in image\n");
			exit(1);
		}
		if (!rewrite(image, AO_RADIO_CAL, cal_int, sizeof (cal_int))) {
			fprintf(stderr, "Cannot rewrite radio calibration at %04x\n", AO_RADIO_CAL);
			exit(1);
		}
	}
	if (!tty)
		tty = cc_usbdevs_find_by_arg(device, "TIDongle");
	dbg = ccdbg_open(tty);
	if (!dbg)
		exit (1);

	ccdbg_add_debug(CC_DEBUG_FLASH);

	ccdbg_debug_mode(dbg);
	ccdbg_halt(dbg);
	if (image->address == 0xf000) {
		printf("Loading %d bytes to execute from RAM\n",
		       image->length);
		ccdbg_write_hex_image(dbg, image, 0);
	} else if (image->address == 0x0000) {
		printf("Loading %d bytes to execute from FLASH\n",
		       image->length);
		ccdbg_flash_hex_image(dbg, image);
	} else {
		printf("Cannot load code to 0x%04x\n",
		       image->address);
		ccdbg_hex_image_free(image);
		ccdbg_close(dbg);
		exit(1);
	}
	ccdbg_set_pc(dbg, image->address);
	ccdbg_resume(dbg);
	ccdbg_close(dbg);
	exit (0);
}
